dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java

```diff
@@ -171,36 +180,103 @@ private static String[] getArgNameArray(AnnotatedMethod<?> method, Map<String, A
             throw new GenericArgumentException(genericParams);
         }
         return result.isEmpty() ? Collections.emptyMap() : result;
     }
 
+    public static Map<String, ArgumentMetadata> getArgumentMap(AnnotatedMethod<?> method) {
+        Map<String, ArgumentMetadata> result = new HashMap<>();
+        for (AnnotatedParameter<?> param : method.getParameters()) {
+            addArgumentMetadata(param, result);
+        }
+        return result.isEmpty() ? Collections.emptyMap() : result;
+    }
+
+    private static void addArgumentMetadata(AnnotatedParameter<?> param, Map<String, ArgumentMetadata> result) {
+        ToolArg argAnnotation = param.getAnnotation(ToolArg.class);
+
+        if (argAnnotation != null) {
+            String argName = resolveArgumentName(param, argAnnotation);
+            boolean isDuplicateArg = result.containsKey(argName);
+
+            result.put(argName, new ArgumentMetadata(param.getBaseType(),
+                                                     param.getPosition(),
+                                                     argAnnotation.description(),
+                                                     argAnnotation.required(),
+                                                     isDuplicateArg));
+        }
+    }
+
+    private static void addArgumentMetadata(AnnotatedParameter<?> param, Map<String, ArgumentMetadata> result, Map<TypeVariable<?>, Type> genericMap) {
+        ToolArg argAnnotation = param.getAnnotation(ToolArg.class);
+
+        if (argAnnotation != null) {
+            String argName = resolveArgumentName(param, argAnnotation);
+            boolean isDuplicateArg = result.containsKey(argName);
+
+            result.put(argName, new ArgumentMetadata(TypeUtility.createCustomType(param.getBaseType(), genericMap),
+                                                     param.getPosition(),
+                                                     argAnnotation.description(),
+                                                     argAnnotation.required(),
+                                                     isDuplicateArg));
+        }
+
+    }
+
+    private static GenericParameterDescription checkConcreteGeneric(Type baseType, Map<TypeVariable<?>, Type> genericMap, ArrayList<String> genericParams,
+                                                                    GenericParameterDescription genericParamDescription) {
+
+        List<Type> genericTypes;
+        if (baseType instanceof ParameterizedType pt) {
+            genericTypes = List.of(pt.getActualTypeArguments());
+
+        } else if (baseType instanceof GenericArrayType gat) {
+            genericTypes = List.of(gat.getGenericComponentType());
+
+        } else if (baseType instanceof TypeVariable<?> tv) {
+            genericTypes = List.of(tv);
+
+        } else if (baseType.getClass().isArray()) {
+            Type elementType = ((Class<?>) baseType).getComponentType();
+            genericTypes = List.of(elementType);
```

Comment:
```suggestion
        } else if (baseType instanceof Class clazz && clazz.isArray()) {
            Type elementType = clazz.getComponentType();
            genericTypes = List.of(elementType);
```

Should be this I think? `baseType.getClass()` would return one of the subtypes of `Type`, whereas I think you want `baseType` as a `Class`